# Asset-based lending: "a primer in python"<br>
>## <font color = green>Asset_based lending (ABL) is an efficient and flexible technique to finance working capital heavy businesses.  It works quite well for noninvestment grade manufacturers and distributors that want to finance inventories and accounts.  In this notebook I've written a simple ABL platform in python to illustrate ABL's capabilities.</font>

>__[ABL from Wikipedia](https://en.wikipedia.org/wiki/Asset-based_lending)__




### Step one.  import our dependencies

In [145]:
from datetime import date
import pandas as pd
import matplotlib.pyplot as plt


In [146]:
from loan_sys_objects import Customer, Loan, Accounts_receivable, Interest
from loan_sys_functions import excess_availability, collections, loan_advance, daily_build
from loan_sys_functions import loan_hist, date_hist, i_rate_hist, accrued_interest_hist, liquidity_hist

### Step two.  Create an instance of the following:
* Customer object
* Loan object
* Accounts_receivable object
* Interest object


In [147]:
ABC_mfg = Customer('ABC_mfg', 'abc@gmail.com', 'Atlanta')
ABC_loan = Loan(5000000, 0)
ABC_collateral = Accounts_receivable(6000000, 600000, .85)
ABC_interest = Interest(.03, .0325)

#### Verify the customer object:

In [148]:
ABC_mfg

Customer name: ABC_mfg
Customer email: abc@gmail.com
Customer Location: Atlanta 

#### Verify the loan object:

In [149]:
ABC_loan

Commitment:$5,000,000
Loan: $0

#### Verify the collateral object:

In [150]:
ABC_collateral

Total Aging:$6,000,000
Past Dues: $600,000
Advance Rate: 85.00%

#### Verify the interest object

In [151]:
ABC_interest

Base Rate:3.00%
Margin: 3.25%

In [152]:
f'The interest rate is the sum of the base rate plus the margin: {ABC_interest.interest_rate():.2%}'

'The interest rate is the sum of the base rate plus the margin: 6.25%'

### <font color = green > To summarize:
* ABC mfg has a  commited credit facility secured by the company's accounts receivable.  The initial loan balance is $0.
* ABC mfg has pledged 6,000,000 of accounts to the lender but the lender but determined that past due accounts are 600,000 and will lend 85% of the eligible accounts.
* The interest rate consists of two parts, a base rate and a margin.  The base rate is typically tied to an index such as 90 libor, US prime rate or some similar market rate.  The number is variable, that is it can fluctuate with the market.  The lenders margin is 3.25% above the index.  This percentage is contractually fixed.</font>



> #### An important concept is the term "availability".  For many borrowers excess availability is synonomous with liquidity

In [153]:
collateral_availability = ABC_collateral.loan_availability()
f'Collateral Availability totals ${collateral_availability:,.2f}.  This equates to gross acounts minus past dues and subsequently reduced by the advance rate of 85%'

'Collateral Availability totals $4,590,000.00.  This equates to gross acounts minus past dues and subsequently reduced by the advance rate of 85%'

In [154]:
excess_avail = excess_availability(ABC_collateral, ABC_loan)
f'Excess Availability totals ${excess_avail:,.2f}.  Excess availability equates to collateral availability less the loan balance'

'Excess Availability totals $4,590,000.00.  Excess availability equates to collateral availability less the loan balance'

### So far, excess Availability is the same as total Availability that's true because the borrower has a zero loan balance.  Lets borrow some money and see what happens

#### We will call the loan_draw method on the loan object to make a $50,000 loan advance

In [155]:
f"The borrower draws down $50,000 and the new loan balance is ${ABC_loan.loan_draw(50000):,.2f}"

'The borrower draws down $50,000 and the new loan balance is $50,000.00'

In [156]:
f" After the draw the excess availability has dropped $50,000 to ${excess_availability(ABC_collateral, ABC_loan):,.2f}"

' After the draw the excess availability has dropped $50,000 to $4,540,000.00'

In [157]:
ABC_loan

Commitment:$5,000,000
Loan: $50,000

In [158]:
ABC_collateral

Total Aging:$6,000,000
Past Dues: $600,000
Advance Rate: 85.00%

#### Next the borrower collects $10,000 of accounts.  We apply the collections to reduce the loan balance and to reduce the collateral balance

In [159]:
collections(ABC_collateral, ABC_loan, 10000)

In [160]:
f"The new loan amount is ${ABC_loan.loan:,.2f}" 

'The new loan amount is $40,000.00'

In [161]:
f"The new excess availability is ${excess_availability(ABC_collateral, ABC_loan):,.2f}"

'The new excess availability is $4,541,500.00'

#### Let's recap:

In [162]:
ABC_loan

Commitment:$5,000,000
Loan: $40,000

In [163]:
ABC_collateral

Total Aging:$5,990,000
Past Dues: $600,000
Advance Rate: 85.00%

In [164]:
excess_avail = excess_availability(ABC_collateral, ABC_loan)
f'Excess Availability totals ${excess_avail:,.2f}.'

'Excess Availability totals $4,541,500.00.'

> ### <font color = 'green'> Once again to summarize:
* The borrower opened with no loan and 4,590,000 in excess undrawn availability on the line of credit.
* The borrower then took a draw for 50,000.  The loan went up and excess availability went down, both by 50,000
* Next the borrower collected 10,000 of accounts receivable. The loan decreases by $10,000 and the collateral decreases by $10,000 but the excess availability, the liquidity of the business increases by 1,500.

In [178]:
# date_hist = []
# loan_hist = []
# i_rate_hist = []
# accrued_interest_hist = []

In [166]:
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,1,ABC_interest)
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,2,ABC_interest)
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,3,ABC_interest)
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,4,ABC_interest)
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,5,ABC_interest)
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,6,ABC_interest)
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,7,ABC_interest)
loan_advance(ABC_collateral,ABC_loan,500000,2019,1,8,ABC_interest)


In [167]:
ABC_loan.loan

4040000

In [168]:
loan_hist

[0,
 540000,
 1040000,
 1540000,
 2040000,
 2540000,
 3040000,
 3540000,
 4040000,
 540000,
 1040000,
 1540000,
 2040000,
 2540000,
 3040000,
 3540000,
 4040000,
 540000,
 1040000,
 1540000,
 2040000,
 2540000,
 3040000,
 3540000,
 4040000,
 540000,
 1040000,
 1540000,
 2040000,
 2540000,
 3040000,
 3540000,
 4040000,
 540000,
 1040000,
 1540000,
 2040000,
 2540000,
 3040000,
 3540000,
 4040000]

In [169]:
date_hist

[datetime.date(2018, 12, 31),
 datetime.date(2019, 1, 1),
 datetime.date(2019, 1, 2),
 datetime.date(2019, 1, 3),
 datetime.date(2019, 1, 4),
 datetime.date(2019, 1, 5),
 datetime.date(2019, 1, 6),
 datetime.date(2019, 1, 7),
 datetime.date(2019, 1, 8),
 datetime.date(2019, 1, 1),
 datetime.date(2019, 1, 2),
 datetime.date(2019, 1, 3),
 datetime.date(2019, 1, 4),
 datetime.date(2019, 1, 5),
 datetime.date(2019, 1, 6),
 datetime.date(2019, 1, 7),
 datetime.date(2019, 1, 8),
 datetime.date(2019, 1, 1),
 datetime.date(2019, 1, 2),
 datetime.date(2019, 1, 3),
 datetime.date(2019, 1, 4),
 datetime.date(2019, 1, 5),
 datetime.date(2019, 1, 6),
 datetime.date(2019, 1, 7),
 datetime.date(2019, 1, 8),
 datetime.date(2019, 1, 1),
 datetime.date(2019, 1, 2),
 datetime.date(2019, 1, 3),
 datetime.date(2019, 1, 4),
 datetime.date(2019, 1, 5),
 datetime.date(2019, 1, 6),
 datetime.date(2019, 1, 7),
 datetime.date(2019, 1, 8),
 datetime.date(2019, 1, 1),
 datetime.date(2019, 1, 2),
 datetime.date(201

In [170]:
i_rate_hist

[0,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625,
 0.0625]

In [171]:
accrued_interest_hist

[0,
 93.75,
 180.55555555555554,
 267.3611111111111,
 354.1666666666667,
 440.97222222222223,
 527.7777777777778,
 614.5833333333334,
 701.3888888888889,
 93.75,
 180.55555555555554,
 267.3611111111111,
 354.1666666666667,
 440.97222222222223,
 527.7777777777778,
 614.5833333333334,
 701.3888888888889,
 93.75,
 180.55555555555554,
 267.3611111111111,
 354.1666666666667,
 440.97222222222223,
 527.7777777777778,
 614.5833333333334,
 701.3888888888889,
 93.75,
 180.55555555555554,
 267.3611111111111,
 354.1666666666667,
 440.97222222222223,
 527.7777777777778,
 614.5833333333334,
 701.3888888888889,
 93.75,
 180.55555555555554,
 267.3611111111111,
 354.1666666666667,
 440.97222222222223,
 527.7777777777778,
 614.5833333333334,
 701.3888888888889]

In [172]:
liquidity_hist

[0,
 4041500.0,
 3541500.0,
 3041500.0,
 2541500.0,
 2041500.0,
 1541500.0,
 1041500.0,
 541500.0,
 4041500.0,
 3541500.0,
 3041500.0,
 2541500.0,
 2041500.0,
 1541500.0,
 1041500.0,
 541500.0,
 4041500.0,
 3541500.0,
 3041500.0,
 2541500.0,
 2041500.0,
 1541500.0,
 1041500.0,
 541500.0,
 4041500.0,
 3541500.0,
 3041500.0,
 2541500.0,
 2041500.0,
 1541500.0,
 1041500.0,
 541500.0,
 4041500.0,
 3541500.0,
 3041500.0,
 2541500.0,
 2041500.0,
 1541500.0,
 1041500.0,
 541500.0]

In [173]:
len(date_hist), len(loan_hist), len(i_rate_hist), len(accrued_interest_hist)
# date_hist_str = [day.strftime for day in date_hist]
# date_hist_str = [i.strftime('%b-%d') for i in date_hist]
date_hist_str = [i.strftime('%b-%d') for i in date_hist]

In [174]:
date_hist_str

['Dec-31',
 'Jan-01',
 'Jan-02',
 'Jan-03',
 'Jan-04',
 'Jan-05',
 'Jan-06',
 'Jan-07',
 'Jan-08',
 'Jan-01',
 'Jan-02',
 'Jan-03',
 'Jan-04',
 'Jan-05',
 'Jan-06',
 'Jan-07',
 'Jan-08',
 'Jan-01',
 'Jan-02',
 'Jan-03',
 'Jan-04',
 'Jan-05',
 'Jan-06',
 'Jan-07',
 'Jan-08',
 'Jan-01',
 'Jan-02',
 'Jan-03',
 'Jan-04',
 'Jan-05',
 'Jan-06',
 'Jan-07',
 'Jan-08',
 'Jan-01',
 'Jan-02',
 'Jan-03',
 'Jan-04',
 'Jan-05',
 'Jan-06',
 'Jan-07',
 'Jan-08']

In [175]:
# df = pd.DataFrame(
#     {'loan':loan_hist,
#      'Interest Rate': i_rate_hist,
#      'Interest Accrual': accrued_interest_hist,
#      'Liquidity': liquidity_hist
#     }, index = date_hist)

df = pd.DataFrame(
    {'loan':loan_hist,
     'Interest Rate': i_rate_hist,
     'Interest Accrual': accrued_interest_hist,
     'Liquidity': liquidity_hist,
     'Date': date_hist_str
    })

In [176]:
df

Date  Interest Accrual  Interest Rate  Liquidity     loan
0   Dec-31          0.000000         0.0000        0.0        0
1   Jan-01         93.750000         0.0625  4041500.0   540000
2   Jan-02        180.555556         0.0625  3541500.0  1040000
3   Jan-03        267.361111         0.0625  3041500.0  1540000
4   Jan-04        354.166667         0.0625  2541500.0  2040000
5   Jan-05        440.972222         0.0625  2041500.0  2540000
6   Jan-06        527.777778         0.0625  1541500.0  3040000
7   Jan-07        614.583333         0.0625  1041500.0  3540000
8   Jan-08        701.388889         0.0625   541500.0  4040000
9   Jan-01         93.750000         0.0625  4041500.0   540000
10  Jan-02        180.555556         0.0625  3541500.0  1040000
11  Jan-03        267.361111         0.0625  3041500.0  1540000
12  Jan-04        354.166667         0.0625  2541500.0  2040000
13  Jan-05        440.972222         0.0625  2041500.0  2540000
14  Jan-06        527.777778         0.0625  1541500.0  3040000
15  Jan-07        614.583333         0.0625  1041500.0  3540000
16  Jan-08        701.388889         0.0625   541500.0  4040000
17  Jan-01         93.750000         0.0625  4041500.0   540000
18  Jan-02        180.555556         0.0625  3541500.0  1040000
19  Jan-03        267.361111         0.0625  3041500.0  1540000
20  Jan-04        354.166667         0.0625  2541500.0  2040000
21  Jan-05        440.972222         0.0625  2041500.0  2540000
22  Jan-06        527.777778         0.0625  1541500.0  3040000
23  Jan-07        614.583333         0.0625  1041500.0  3540000
24  Jan-08        701.388889         0.0625   541500.0  4040000
25  Jan-01         93.750000         0.0625  4041500.0   540000
26  Jan-02        180.555556         0.0625  3541500.0  1040000
27  Jan-03        267.361111         0.0625  3041500.0  1540000
28  Jan-04        354.166667         0.0625  2541500.0  2040000
29  Jan-05        440.972222         0.0625  2041500.0  2540000
30  Jan-06        527.777778         0.0625  1541500.0  3040000
31  Jan-07        614.583333         0.0625  1041500.0  3540000
32  Jan-08        701.388889         0.0625   541500.0  4040000
33  Jan-01         93.750000         0.0625  4041500.0   540000
34  Jan-02        180.555556         0.0625  3541500.0  1040000
35  Jan-03        267.361111         0.0625  3041500.0  1540000
36  Jan-04        354.166667         0.0625  2541500.0  2040000
37  Jan-05        440.972222         0.0625  2041500.0  2540000
38  Jan-06        527.777778         0.0625  1541500.0  3040000
39  Jan-07        614.583333         0.0625  1041500.0  3540000
40  Jan-08        701.388889         0.0625   541500.0  4040000

In [177]:
%matplotlib inline
# plt.scatter(loan_hist, date_hist)
df.plot(kind='scatter',x='Date',y='Liquidity',color='red')
plt.show()

ValueError: scatter requires x column to be numeric

In [ ]:
ABC_loan

In [ ]:
excess_availability(ABC_collateral, ABC_loan)